drivers/ethernet/eth_stm32_hal.c

```diff
@@ -256,14 +256,16 @@ static inline void setup_mac_filter(ETH_HandleTypeDef *heth)
 
 	/* disable multicast perfect filtering */
 	tmp &= ~(ETH_MULTICASTFRAMESFILTER_PERFECTHASHTABLE |
-#if !defined(CONFIG_ETH_STM32_MULTICAST_FILTER)
 		 ETH_MULTICASTFRAMESFILTER_HASHTABLE |
-#endif /* CONFIG_ETH_STM32_MULTICAST_FILTER */
-		 ETH_MULTICASTFRAMESFILTER_PERFECT);
+		 ETH_MULTICASTFRAMESFILTER_PERFECT |
+		 ETH_MULTICASTFRAMESFILTER_NONE);
 
 #if defined(CONFIG_ETH_STM32_MULTICAST_FILTER)
 	/* enable multicast hash receive filter */
 	tmp |= ETH_MULTICASTFRAMESFILTER_HASHTABLE;
+#else
+	/* enable receiving all multicast frames */
+	tmp |= ETH_MULTICASTFRAMESFILTER_NONE;
 #endif /* CONFIG_ETH_STM32_MULTICAST_FILTER */
```

Comment:
For readability, consider use of `IS_ENABLED`.

```suggestion
	if (IS_ENABLED(CONFIG_ETH_STM32_MULTICAST_FILTER)) {
		/* enable multicast hash receive filter */
		tmp |= ETH_MULTICASTFRAMESFILTER_HASHTABLE;
	} else {
		/* enable receiving all multicast frames */
		tmp |= ETH_MULTICASTFRAMESFILTER_NONE;
	}
```

Reply:
Done, thanks!